### PR TITLE
[algo] feat: SAC Performance Improvements for Pi0.5

### PR DESCRIPTION
**Pull request by Miical**
### What does this PR do?

This PR fixes several issues that affected the performance of SAC when training the pi0.5 model, and further improves the overall algorithm performance of SAC.  
At the current stage, we are able to achieve approximately **90% success rate** on the **libero-spatial** task set.

---

#### New Features

1. Introduced the **Flow-SDE method** during the rollout phase to enhance action exploration.
2. Upgrade the critic network, improving evaluation accuracy.

---

#### Bug Fixes

1. Updated compatibility with the latest codebase.
2. Synchronized parameters with the critic head during **target network initialization**.
3. Fixed a bug in **action regularization**.
4. Fixed parameter misalignment during **target network updates**.
5. Fixed an issue where loading a smaller replay buffer checkpoint caused incorrect indexing.
6. Refactored `sac_ray_trainer` for better readability and fixed the `validate` functionality.
7. Fixed the `prepare_libero_dataset` script so it can generate all task sets in LIBERO.

---

#### Tunable Parameters & Metrics

1. Added **Q-value display** of the current action in rendered videos.
2. Improved the visual quality of text overlays in videos.
3. Added a parameter to adjust the ratio between **rollout and training steps**.
4. Added parameters to control the update frequency of **actor and critic**.
5. Introduced a separate optimizer for the critic head, allowing independent learning rate tuning.
6. Added metrics for **positive/negative sample Q-values**, as well as their difference.

---

#### Current Results

At this stage, we achieve strong performance on single-task settings.  
In multi-task scenarios, we reach approximately **90% success rate** on the **libero-spatial benchmark**.
Below are some example cases we have previously tested:

| Benchmark                  | Setting       | Success Rate        | Improvement |
|---------------------------|--------------|---------------------|-------------|
| Libero-Spatial            | Single Task  | ~64% → ~99%         | ~1.55×      |
| Libero-10                 | Single Task  | ~15% → ~99%         | ~6.60×      |
| Libero-Spatial (Test Suite) | Multi Task | ~68% → ~90%         | ~1.32×      |

Libero Spatial Single Task
corner case

Libero-10 Single Task
libero-10

Libero-Spatial Multi Task
截屏2026-03-18 14 35 02

---

#### Future Work

We will continue exploring ways to further improve performance and extend to **real-world reinforcement learning**.  
If you are interested, feel free to contact us.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

Before starting, please prepare the training and validation datasets using the following script:

```bash
python3 prepare_libero_dataset.py
```

We recommend using the **disaggregated script** to run this PR:

```bash
sh ./verl/experimental/vla/run_pi05_libero_sac_disagg.sh
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/experimental/vla/models/pi0_torch/modeling_pi0_torch.py [hunk 1]
```diff
@@ -458,6 +581,17 @@ def sac_forward_state_features(
     @override
     @torch.no_grad()
     def sac_update_target_network(self, tau: float):
-        for target_head, head in zip(self.target_network_heads, self.critic_heads, strict=False):
-            for target_param, param in zip(target_head.parameters(), head.parameters(), strict=False):
-                target_param.data.mul_(1.0 - tau).add_(param.data, alpha=tau)
```
[the remark this answers is not in the document]
Comment (Miical) [reply verl/experimental/vla/models/pi0_torch/modeling_pi0_torch.py:463]: > high
> 
> The target network update relies on `zip`ping parameter iterators with `strict=False`. This is fragile because the order of parameters is not guaranteed to be identical between the source and target networks, especially with complex models or framework updates. This could lead to incorrect or silent weight transfer failures. The robust way to perform Polyak averaging is by using `state_dict`, which matches parameters by name.

Here, zip is used to combine lists. The internal structure is in the form of state_dict. There is no issue here.

File: verl/experimental/vla/models/modules/mlp.py [hunk 2]
```diff
@@ -57,54 +60,64 @@ def __init__(
...
+        # (works because only the last Linear has out_features == self.output_dim in this MLP)
+        is_output_layer = m.out_features == self.output_dim
+
+        if is_output_layer:
+            init.uniform_(m.weight, -self.output_init_scale, self.output_init_scale)
+        else:
             if self.init_method == "kaiming":
-                # Use 'relu' as default nonlinearity for Kaiming
-                nonlinearity = self.activation_name if self.activation_name in ["relu", "leaky_relu"] else "relu"
-                init.kaiming_normal_(m.weight, nonlinearity=nonlinearity)
```
[the remark this answers is not in the document]
Comment (Miical) [reply verl/experimental/vla/models/modules/mlp.py:96]: This area has been set to 0.2.

File: verl/experimental/vla/sac/sac_actor.py [hunk 3]
```diff
@@ -283,76 +383,133 @@ def _forward_actor(self, micro_batch: TensorDict) -> tuple[torch.Tensor, torch.T
...
+                    is_first_micro_batch=(batch_idx == 0),
+                )
                 (raw_actor_loss / grad_accum_steps).backward()
                 actor_loss_list.append(raw_actor_loss.detach().item())
-                actor_logprobs_list.append(log_probs.detach())
+                if log_probs is not None:
+                    actor_logprobs_list.append(log_probs.detach())
+                actor_qvalues_list.append(q_values.detach())
+                actor_forward_metrics.update(actor_forward_metrics_mb)
             actor_grad_norm = self._optimizer_step()
```
[the remark this answers is not in the document]
Comment (Miical) [reply verl/experimental/vla/sac/sac_actor.py:345]: This is to be compatible with the previous designs.
When updating the actor, we freeze the gradients of the critic.
The update of the critic is handled by a separate optimizer.

File: verl/experimental/vla/sac/sac_actor.py [hunk 4]
```diff
@@ -223,26 +302,39 @@ def _calculate_critic_loss(
         alpha = self._get_alpha()
 
         with torch.no_grad():
-            y = rewards + valid * gamma * (q_target - alpha * next_log_prob)
```
[the remark this answers is not in the document]
Comment (Miical) [reply verl/experimental/vla/sac/sac_actor.py:226]: There is no problem here. It has been fixed in the new version of the code.
